src/query/api/v1/handler/prometheus/debugger/validate.go

```diff
@@ -0,0 +1,266 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package debug
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"math"
+	"net/http"
+	"time"
+
+	"github.com/m3db/m3/src/query/api/v1/handler"
+	"github.com/m3db/m3/src/query/api/v1/handler/prometheus"
+	"github.com/m3db/m3/src/query/api/v1/handler/prometheus/native"
+	"github.com/m3db/m3/src/query/executor"
+	"github.com/m3db/m3/src/query/models"
+	"github.com/m3db/m3/src/query/storage/debugger"
+	"github.com/m3db/m3/src/query/ts"
+	qjson "github.com/m3db/m3/src/query/util/json"
+	"github.com/m3db/m3/src/query/util/logging"
+	"github.com/m3db/m3/src/x/net/http"
+
+	"github.com/uber-go/tally"
+	"go.uber.org/zap"
+)
+
+const (
+	// PromDebugURL is the url for the Prom and m3query debugging tool
+	PromDebugURL = handler.RoutePrefixV1 + "/debug/validate_query"
+
+	// PromDebugHTTPMethod is the HTTP method used with this resource.
+	PromDebugHTTPMethod = http.MethodPost
+)
+
+// PromDebugHandler represents a handler for prometheus debug endpoint.
+type PromDebugHandler struct {
+	engine      *executor.Engine
+	scope       tally.Scope
+	readHandler *native.PromReadHandler
+}
+
+// NewPromDebugHandler returns a new instance of handler.
+func NewPromDebugHandler(
+	h *native.PromReadHandler,
+	scope tally.Scope,
+) *PromDebugHandler {
+	return &PromDebugHandler{
+		scope:       scope,
+		readHandler: h,
+	}
+}
+
+func (h *PromDebugHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	ctx := context.WithValue(r.Context(), handler.HeaderKey, r.Header)
+	logger := logging.WithContext(ctx)
+
+	data, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		logger.Error("unable to read data", zap.Error(err))
+		xhttp.Error(w, err, http.StatusBadRequest)
+		return
+	}
+
+	var promDebug prometheus.PromDebug
+	if err := json.Unmarshal(data, &promDebug); err != nil {
+		logger.Error("unable to unmarshal data", zap.Error(err))
+		xhttp.Error(w, err, http.StatusBadRequest)
+		return
+	}
+
+	s, err := debug.NewStorage(promDebug.Input)
+	if err != nil {
+		logger.Error("unable to create storage", zap.Error(err))
+		xhttp.Error(w, err, http.StatusBadRequest)
+		return
+	}
+
+	h.engine = executor.NewEngine(s, h.scope.SubScope("debug_engine"))
```

Comment:
Should this be creating a new engine each time you run?

Reply:
Yeah I discussed this with @nikunjgit and it's fine since it's a debug endpoint